contracts/cw1-subkeys/src/contract.rs

```diff
@@ -507,477 +614,388 @@ mod tests {
+                    ..Spender::default()
+                },
+            ],
+            ..SuiteConfig::default()
+        }
+        .init();
 
-        let info = mock_info(owner, &[]);
-        setup_test_case(
-            deps.as_mut(),
-            &info,
-            &admins,
-            &initial_spenders,
-            &initial_allowances,
-            &initial_expirations,
-        );
+        // let's try pagination.
+        //
+        // Check is tricky, as there is no guarantee about order expiration are received (as it is
+        // dependent at least on ordering of insertions), so to check if pagination works, all what
+        // can we do is to ensure parts are of expected size, and that collectively all allowances
+        // are returned.
+        let batch1 = query_all_allowances(deps.as_ref(), None, Some(2))
+            .unwrap()
+            .allowances;
+        assert_eq!(2, batch1.len());
 
-        // let's try pagination
-        let allowances = query_all_allowances(deps.as_ref(), None, Some(2))
+        // now continue from after the last one
+        let batch2 = query_all_allowances(deps.as_ref(), Some(batch1[1].spender.clone()), Some(2))
             .unwrap()
             .allowances;
-        assert_eq!(2, allowances.len());
-        assert_eq!(
-            allowances[0],
+        assert_eq!(1, batch2.len());
+
+        let expected = vec![
             AllowanceInfo {
-                spender: spender1.into(),
-                balance: NativeBalance(initial_allowances.clone()),
+                spender: SPENDER1.to_owned(),
+                balance: NativeBalance(vec![coin(1234, TOKEN)]),
                 expires: Expiration::Never {},
-            }
-        );
-        assert_eq!(
-            allowances[1],
+            },
             AllowanceInfo {
-                spender: spender2.to_string(),
-                balance: NativeBalance(initial_allowances.clone()),
+                spender: SPENDER2.to_owned(),
+                balance: NativeBalance(vec![coin(2345, TOKEN)]),
                 expires: Expiration::Never {},
-            }
-        );
-
-        // now continue from after the last one
-        let allowances = query_all_allowances(deps.as_ref(), Some(spender2.into()), Some(2))
-            .unwrap()
-            .allowances;
-        assert_eq!(1, allowances.len());
-        assert_eq!(
-            allowances[0],
+            },
             AllowanceInfo {
-                spender: spender3.into(),
-                balance: NativeBalance(initial_allowances),
-                expires: expires_later,
-            }
+                spender: SPENDER3.to_owned(),
+                balance: NativeBalance(vec![coin(3456, TOKEN)]),
+                expires: Expiration::AtHeight(12345),
+            },
+        ];
+
+        assert_sorted_eq!(
+            expected,
+            [batch1, batch2].concat(),
+            |l: &AllowanceInfo, r: &AllowanceInfo| l.spender.cmp(&r.spender)
```

Comment:
Nice trick relying on `PartialEq` for sorting, instead of requiring `Ord` here.

I understand it would make sense to just add `Ord` (and `Eq`) to `AllowanceInfo`, and simplify this.

As that can require adding them to other stuff, including some structs in cosmwasm-std, I would say, let's create an issue to do just this, at least for cosmwasm-std.

Reply:
Defienetely implementing `Ord` for `AllowanceInfo` does not make sense - `Ord` is for types with reasonable sorting, and I find implementing it for this kind of type a missuse. Note that even standard library doesn't implement sort for types like floats (even if you could technically provide implementation which would somehow properly sort it). Actually it is questionable if there is proper way to pick proper key for comparison - is this anyhow intuitive that allowance with sender earlier in alphabet is in any way smaller than the other one?

My personal opinion is, that even if this adds some code, it avoids any confusion how things are sorted (and why). I think about two possible simplifications: 
1. try to write macro in the way, that it would be more elision friendly so type hints are not needed (actually type hints are half the length of the closure!) - unfortunatelly I have no idea how to approach this
2. extract the comparison function to `AllowanceInfo`, but not as an implementation of `Ord`, but rather testing purpose method like `AllowanceInfo::cmp_by_spender` - actually I like the idea

Reply:
> 2\. extract the comparison function to `AllowanceInfo`, but not as an implementation of `Ord`, but rather testing purpose method like `AllowanceInfo::cmp_by_spender` - actually I like the idea

Sounds good. That, and implementing these helpers only for tests. That could include deriving something like `Ord`, which does not make sense / adds unnecessary complexity, only for tests, using conditional compilation.